src/activity_actor.h

```diff
@@ -19,6 +20,131 @@ class JsonIn;
 class JsonOut;
 class player_activity;
 
+struct simple_task {
+    // Name of the target that's being processed
+    const std::string target_name;
+    /* Total number of moves required for this target/task to complete */
+    const int moves_total = 0;
+    /* The number of moves remaining for this target/task to complete */
+    int moves_left = 0;
+
+    inline bool complete() const {
+        return moves_left <= 0;
+    }
+
+    //Json stuff
+    void serialize( JsonOut &json ) const;
+};
+
+/*
+ * Special class to track progress of current activity
+ * Outside is a queue with slighly altered functionality
+*/
+class progress_counter
+{
+    private:
+        /** Total number of moves required to complete the activity aka all the tasks */
+        int moves_total = 0;
+        /** The number of moves remaining in this activity before it is complete aka all the tasks */
+        int moves_left = 0;
+        //Index of current task - 1-based
+        int idx = 1;
+        //Counts total amount of tasks - done and in queue
+        int total_tasks = 0;
+
+        std::deque<simple_task> targets;
+
+    public:
+        inline void emplace( std::string name, int moves_total_ ) {
+            moves_total += moves_total_;
+            moves_left += moves_total_;
+            targets.emplace_back( simple_task {
+                .target_name = name,
+                .moves_total = moves_total_,
+                .moves_left = moves_total_
+            } );
+            total_tasks++;
+        }
+        inline void emplace( std::string name, int moves_total_, int moves_left_ ) {
+            moves_total += moves_total_;
+            moves_left += moves_left_;
+            targets.emplace_back( simple_task {
+                .target_name = name,
+                .moves_total = moves_total_,
+                .moves_left = moves_left_
+            } );
+            total_tasks++;
+        }
+        inline void pop() {
+            if( targets.empty() ) {
+                debugmsg( "task was popped out of empty progress queue" );
+                return;
+            }
+            moves_left -= targets.front().moves_left;
+            targets.pop_front();
+            idx++;
+        }
+        inline void purge() {
+            if( targets.empty() ) {
+                debugmsg( "task was purged out of empty progress queue" );
+                return;
+            }
```

Comment:
```suggestion
            if( empty() ) {
                return;
            }
```
Same reasoning as above

Reply:
resolution from discord discussion:
leaving the debugmsg as is for now so we can see if it ends up generating bug reports. If not then it can have the severity lowered from `DL::Error` to `DL::Debug`